### PR TITLE
Some `Symbol` related improvements

### PR DESCRIPTION
**Pull request by nnethercote**
These commits make things nicer and avoid some `Symbol::as_str()` calls.

r? @oli-obk

### CODE REVIEW

File: src/librustc_builtin_macros/deriving/generic/ty.rs [hunk 1]
```diff
@@ -225,14 +225,13 @@ fn mk_generics(params: Vec<ast::GenericParam>, span: Span) -> Generics {
 
 /// Lifetimes and bounds on type parameters
```
Comment (oli-obk) [inline src/librustc_builtin_macros/deriving/generic/ty.rs:226]: This documentation comment is now outdated

**Comment by oli-obk** [comment]
r=me with documentation nit resolved.

**Comment by nnethercote** [comment]
@bors r=oli-obk, rollup=iffy

Iffy because it's fairly large.

**Comment by nnethercote** [comment]
I don't think this will have a perf effect, but let's check to be sure.

@bors try @rust-timer queue

**Comment by oli-obk** [comment]
perf looks clean with potential small improvements, just needs a rebase

**Comment by nnethercote** [comment]
I rebase.

@bors r=oli-obk

**Pull request merged**